### ISSUE TITLE
Core: Display readonly input component as text

### ISSUE THREAD

**Issue by Rapster** [primefaces/primefaces#4278]
In a lot of web application, data displayed as input can also be displaying as output. For exemple, creation of a user and detail of a user. Basically, it'll show the same info but only the components change: input (creation) -> output (detail). In our company, we have a bean called `mode` which has 2 values: READ/UPDATE. Depending the mode set, we'll display the input or the ouput. For that, we created a bunch of composite which basically does that:

```xml
<ui:component>
    <cc:implementation>
        <p:input value="#{bean.value}" rendered="#{mode.value == 'UPDATE'}">
        <h:outputText value="#{bean.value}" rendered="#{mode.value == 'READ'}">
    </cc:implementation>
</ui:component>
```

Instead I'd like to be able to write this:
```xml
<p:input value="#{bean.value}" readonly="#{mode.value == 'READ'}">
```

First, the writing is a lot more simple and second, you don't congest the JSF tree with "unused" component or painful composite. This behavior could be set via a context-param (just like RTL). WDYT?

Components I have in mind so far are more or less all the input components

Add options:

- separator for multiple values (default ", ")
- default empty value (default is empty, but could be: "none", "non renseigné" etc.)
- Add PrimeFaces#setReadOnly(String clientId, boolean reaodnly): set readonly to all clientId children and update it/them

**Comment by OBreidenbach** [comment]
👍 We do pretty much the same. 
I am just not sure if there is always a generally valid read only view for all the input components.

**Comment by Rapster** [comment]
Yes I know, this is why I said "more or less all the input components". My thougths are, for:
- multiple values (like PickList, Chips, Autocomplete etc.): display them with a separator (e.g Berlin, Paris, London)
- simple value: simple text (e.g Berlin)

Though for Rating, we should keep it as it is. ColorPicker could be a simple box filled with a color or rgb code? I'd like to implement that from InputRenderer (but not sure if it'll be possible, it's okay I have something else in mind)

**Comment by tandraschko** [comment]
Basically i like the idea, however we should start the discussion after 6.3.
It could even be possible with just CSS (also its not the nicest way).

**Comment by Rapster** [comment]
I started to work on something, I'll submit a PR to exchange ideas about this. CSS sounds like a lot of fiddling and it'll be a lot more easier to implement server side

**Comment by Rapster** [comment]
I just made a proposal (I didn't test everything, but just to give an idea of what it could look like). I only focused on InputRenderer kinda component (but that should be enough no?).

Basically, the idea is prety simple, depending the kind of component (select, simple input, tricky input like autocomplete/picklist) have their own way to retrieve the "readony value": see [ReadOnlyDecoder](https://github.com/Rapster/primefaces/blob/526a638d29aff4ced1596196f329a0bd7d1e4cd0/src/main/java/org/primefaces/component/api/ReadOnlyDecoder.java).

Because the whole encoding of the readonly value happen during UIComponent#encodeEnd method, I "interfaced" InputRenderer with two well known methods: encodeMarkup and encodeScript. (Have a look from here https://github.com/Rapster/primefaces/blob/526a638d29aff4ced1596196f329a0bd7d1e4cd0/src/main/java/org/primefaces/renderkit/InputRenderer.java#L44) Most of our component implement these, or if not, should :smile: 

But the contract set by these methods are not fully respected by 3 components (inheriting InputRenderer): CalendarRenderer, InputNumberRenderer and ColorPickerRenderer. Methods have additionnal parameters so they can be passed to encodeMarkup and encodeScript. This is where I struggled...  I need to pass additional parameters to these 2 methods, but should be computed only once (for perf reasons). I had a few ideas but I'd like to hear it from you. 

I try to make this post not too verbose so I'll stop from now :smile: 
@tandraschko @melloware what do you think?

**Comment by melloware** [comment]
Not bad.  How will user opt-in to using this or not using this?  will it be a web.xml CONTEXT setting?

**Comment by Rapster** [comment]
Yes, there will be an option in web.xml (disabled by default)

**Comment by Rapster** [comment]
@tandraschko @melloware whenever you have some time, I'd like to hear from you on [my comment](https://github.com/primefaces/primefaces/issues/4278#issuecomment-443165936) and specially on this:

> But the contract set by these methods are not fully respected by 3 components (inheriting InputRenderer): CalendarRenderer, InputNumberRenderer and ColorPickerRenderer. Methods have additionnal parameters so they can be passed to encodeMarkup and encodeScript. This is where I struggled... I need to pass additional parameters to these 2 methods, but should be computed only once (for perf reasons).

Let's put some code, it might help. `InputRenderer#encodeEnd` would look like this:
```java
    @Override
    public final void encodeEnd(FacesContext context, UIComponent component) throws IOException {
        PrimeConfiguration config = PrimeApplicationContext.getCurrentInstance(context).getConfig();
        if (readOnlyDecoder != null
                && isReadOnly((UIInput) component)
                && config.isDisplayReadOnlyAsText()) {
            encodeReadOnly(context, component, config);
        } else {
            encodeMarkup(context, (T) component);
            encodeScript(context, (T) component);
        }
    }
```

And to show the problem, let's take `InputNumberRenderer` (extending `InputRenderer`):
```java
    @Override
    public void encodeEnd(FacesContext context, UIComponent component) throws IOException {
        InputNumber inputNumber = (InputNumber) component;

        Object value = inputNumber.getValue();
        String valueToRender = ComponentUtils.getValueToRender(context, inputNumber, value);
        if (valueToRender == null) {
            valueToRender = "";
        }

        encodeMarkup(context, inputNumber, valueToRender);
        encodeScript(context, inputNumber, value, valueToRender);
    }
```

My idea is that, every input renderer in PF should implement encodeMarkup/encodeScript, but `InputNumberRenderer#encodeMarkup` does not follow `InputRenderer#encodeMarkup` signature! This is why in #5356 I asked if it's "bad" to have attributes in renderers, to workaround that problem I could have `value` and `valueToRender` as attributes, but I understand it's a no go that way.

Anyway, let me know your thoughts on that problem I'm facing 😉 

Maybe the easiest way is that, for those renderers, override `InputRenderer#encodeEnd`? But we'll end up with empty encodeMarkup/encodeScript methods...

**Comment by melloware** [comment]
Yep i see the problem.  You need some way of getting valueToRender into the encodeMarkup/Script without having to pass it as a param.  I will defer to @tandraschko as to the safest way to accomplish that.  But other than that I like your proposal.

**Comment by Rapster** [comment]
Or, my "best" option is having something like this:

```
Map<String, Object> parameters = encodeMarkup(context, (T) component);
encodeScript(context, (T) component, parameters);
```

`InputRenderer#encodeMarkup` returns whatever is needed for `InputRenderer#encodeScript`. I think 85% of the component will return `Collections.emptyMap()`

**Comment by melloware** [comment]
Also not a bad idea!

**Comment by tandraschko** [comment]
i will close this for now, its not on our roadmap
